datalad/runner/nonasyncrunner.py

```diff
@@ -35,226 +48,499 @@
-                return
 
-            self.queue.put((self.file.fileno(), data, time.time()))
+class ThreadedRunner:
+    """
+    A class the contains a naive implementation for concurrent sub-process
+    execution. It uses `subprocess.Popen` and threads to read from stdout and
+    stderr of the subprocess, and to write to stdin of the subprocess.
 
+    All read data and timeouts are passed to a protocol instance, which can
+    create the final result.
+    """
+    # Interval in seconds after which we check that a subprocess
+    # is still running.
+    process_check_interval = 0.2
 
-class _StdinWriterThread(threading.Thread):
-    def __init__(self, stdin_data, process, stdin_fileno, q, command=""):
+    def __init__(self,
+                 cmd: Union[str, List],
+                 protocol_class: Type[WitlessProtocol],
+                 stdin: Any,
+                 protocol_kwargs: Optional[Dict] = None,
+                 timeout: Optional[float] = None,
+                 **popen_kwargs
+                 ):
         """
         Parameters
         ----------
-        stdin_data:
-          Data that should be written to the file given by `stdin_fileno´.
-        process:
-          a subprocess.Popen-instance. It is mainly used to access
-          popen._stdin_write(...)
-        q:
-          A queue into which the thread writes a None-data object to
-          indicate that all stdin_data was written.
-        command:
-          The command for which the thread was created. This
-          is mainly used to improve debug output messages.
+        cmd : list or str
+            Command to be executed, passed to `subprocess.Popen`. If cmd
+            is a str, `subprocess.Popen will be called with `shell=True`.
+
+        protocol : WitlessProtocol class or subclass which will be
+            instantiated for managing communication with the subprocess.
+
+            If the protocol is a subclass of
+            `datalad.runner.protocol.GeneratorMixIn`, this function will
+            return a `Generator` which yields whatever the protocol callback
+            fed into `GeneratorMixIn.send_result()`.
+
+            If the protocol is not a subclass of
+            `datalad.runner.protocol.GeneratorMixIn`, the function will return
+            the result created by the protocol method `_generate_result`.
+
+        stdin : file-like, string, bytes, Queue, or None
+            If stdin is a file-like, it will be directly used as stdin for the
+            subprocess. The caller is responsible for writing to it and closing
+            it. If stdin is a string or bytes, those will be fed to stdin of the
+            subprocess. If all data is written, stdin will be closed.
+            If stdin is a Queue, all elements (bytes) put into the Queue will
+            be passed to stdin until None is read from the queue. If None is
+            read, stdin of the subprocess is closed.
```

Comment:
and what happens if value is `None` - `stdin` is used? (docstring doesn't describe that)

Reply:
Description added